File: app/schemas/ai.py
"""Pydantic schemas for the AI Recommendations API."""

from __future__ import annotations

from datetime import date as date_type, datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class AnalyzeRequest(BaseModel):
    department_id: UUID
    date_start: date_type
    date_end: date_type
    provider: str = Field(default="claude", pattern="^(claude|openai|gemini)$")


class AgentResultItem(BaseModel):
    agent_name: str
    title: str
    description: str
    success: bool
    result: Optional[str] = None
    error: Optional[str] = None


class AnalyzeResponse(BaseModel):
    analysis_id: int
    department_id: UUID
    date_start: date_type
    date_end: date_type
    provider: str
    success: bool
    results: dict[str, str]
    errors: dict[str, str]
    skipped: list[str]
    metadata: dict[str, Any]
    created_at: datetime


class HistoryItem(BaseModel):
    id: int
    department_id: UUID
    department_name: Optional[str] = None
    date_start: date_type
    date_end: date_type
    provider: str
    agent_count: int
    created_at: datetime


class PromptInfo(BaseModel):
    agent_name: str
    prompt: str
    source: str  # "db" or "default"
    updated_at: Optional[datetime] = None


class UpdatePromptsRequest(BaseModel):
    prompts: dict[str, str]


class RerunAgentRequest(BaseModel):
    analysis_id: int
    agent_name: str
    new_prompt: Optional[str] = None
    provider: str = Field(default="claude", pattern="^(claude|openai|gemini)$")


class PromptLogItem(BaseModel):
    agent_name: str
    provider: str
    full_prompt: str
    system_prompt: Optional[str] = None
    response_text: Optional[str] = None
    success: bool
    tokens_used: int
    request_timestamp: datetime
    response_timestamp: Optional[datetime] = None
    response_time_seconds: Optional[float] = None
    error_message: Optional[str] = None


class AnalysisDetailResponse(BaseModel):
    id: int
    department_id: UUID
    department_name: Optional[str] = None
    date_start: date_type
    date_end: date_type
    provider: str
    agent_results: dict[str, str]
    created_at: datetime


class AnalysisPromptsResponse(BaseModel):
    analysis: AnalysisDetailResponse
    agents: list[dict]

File: frontend/src/types/ai.ts
export type AIProvider = 'claude' | 'openai' | 'openrouter' | 'gemini'

export interface AIProviderInfo {
  provider: AIProvider
  configured: boolean
  model?: string
  status?: string
  isolated_keys?: Record<string, boolean>
}

export interface AIProvidersResponse {
  supported_providers: AIProvider[]
  default_provider: AIProvider
  providers: Record<AIProvider, AIProviderInfo>
}

export interface AIPromptInfo {
  agent_name: string
  prompt: string
  source: 'db' | 'default'
  updated_at: string | null
}

export interface AIHistoryItem {
  id: number
  department_id: string
  department_name: string | null
  date_start: string
  date_end: string
  provider: AIProvider
  agent_count: number
  created_at: string
}

export interface AIAnalyzeRequest {
  department_id: string
  date_start: string
  date_end: string
  provider: AIProvider
}

export interface AIAnalyzeResponse {
  analysis_id: number
  department_id: string
  date_start: string
  date_end: string
  provider: AIProvider
  success: boolean
  results: Record<string, string>
  errors: Record<string, string>
  skipped: string[]
  metadata: Record<string, unknown>
  created_at: string
}

export interface AIPromptLog {
  full_prompt: string
  system_prompt: string | null
  provider: AIProvider
  success: boolean
  tokens_used: number
  request_timestamp: string
  response_timestamp: string | null
  response_time_seconds: number | null
  error_message: string | null
}

export interface AIAgentDetail {
  agent_name: string
  title: string
  enabled: boolean
  result: string | null
  has_result: boolean
  prompt: AIPromptLog | null
}

export interface AIAnalysisDetail {
  id: number
  department_id: string
  department_name: string | null
  date_start: string
  date_end: string
  provider: AIProvider
  agent_results: Record<string, string>
  created_at: string
}

export interface AIAnalysisPromptsResponse {
  analysis: AIAnalysisDetail
  agents: AIAgentDetail[]
}

export interface AIRerunAgentRequest {
  analysis_id: number
  agent_name: string
  new_prompt?: string
  provider: AIProvider
}

export interface AIRerunAgentResponse {
  success: boolean
  agent_name: string
  result: string | null
  error: string | null
  tokens_used: number
}
